interfaces/repo.go

```diff
@@ -516,3 +516,126 @@ func (r *Repository) securitySnippetsForSnap(snapName string, securitySystem Sec
 	}
 	return snippets, nil
 }
+
+type addSnapError struct {
+	errs []error
+}
+
+func (e *addSnapError) Error() string {
+	return fmt.Sprintf("errors while adding snap: %v", e.errs)
+}
+
+// AddSnap adds plugs and slots declared by the given snap to the repository.
+//
+// This function can be used to implement snap install or, when used along with
+// RemoveSnap, snap upgrade.
+//
+// AddSnap doesn't change existing plugs/slots. The caller is responsible for
+// ensuring that the snap is not present in the repository in any way prior to
+// calling this function. If this constraint is violated then no changes are
+// made and an error is returned.
+//
+// Each added plug/slot is validated according to the corresponding interface.
+// Unknown interfaces and plugs/slots that don't validate are not added.
+// Information about those failures are returned to the caller.
+func (r *Repository) AddSnap(snapInfo *snap.Info) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	snapName := snapInfo.Name
+
+	for plugName := range snapInfo.Plugs {
+		if _, ok := r.plugs[snapName][plugName]; ok {
+			return fmt.Errorf("cannot add plug %s.%s: existing plug in the way", snapName, plugName)
+		}
+	}
+	for slotName := range snapInfo.Slots {
+		if _, ok := r.slots[snapName][slotName]; ok {
+			return fmt.Errorf("cannot add slot %s.%s: existing slot in the way", snapName, slotName)
+		}
+	}
```

Comment:
Nice simplification! Thank you, applied 81f52ebd4057ff89955b3f8f2875dda4fa06e42f